### PR TITLE
Samueleceroni/warehouse/general

### PR DESCRIPTION
**Pull request by samueleceroni**
Completed Implementation of stock and query Stock
Implementation of WarehouseImpl

### CODE REVIEW

File: src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java [hunk 1]
```diff
@@ -0,0 +1,184 @@
...
+
+    private static final String UOM_ERROR = "Quantity unit of measure mismatching.\nBe sure to"
+                                          + "select an article when selecting a quantity filter"
+                                          + " and to select the same unit of measure.";
+    @Nullable
+    private Article article = null;
+    @Nullable
+    private Date expiresBefore = null;
+    @Nullable
+    private Date expiresAfter = null;
```
Comment (Zeegomo) [inline src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java]: this is redundant

File: src/main/java/nwoolcan/model/brewery/warehouse/WarehouseImpl.java [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+    }
+
+    @Override
+    public Result<Empty> addRecord(final Article article, @Nullable final Date expirationDate, final Record record) {
+        return Result.of(new StockImpl(article, expirationDate))
+                     .peek(stock -> updateArticles(article))
+                     .flatMap(stock -> Result.of(this.getStock(stock)))
+                     .flatMap(stock -> stock.addRecord(record))
+                     .toEmpty();
+                     //.flatMap(stock -> Result.ofEmpty());
```
Comment (Zeegomo) [inline src/main/java/nwoolcan/model/brewery/warehouse/WarehouseImpl.java]: useless

Comment (samueleceroni) [reply src/main/java/nwoolcan/model/brewery/warehouse/WarehouseImpl.java]: it is not useless at all. the stock could change in that moment. be aware that the stock is hashed only by certain parameters.

File: src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java (hunk 1 above)
Comment (samueleceroni) [reply src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java]: what are you referring to? ```@Nullable``` is not redundant at all. Try to remove it and you'll see it does not compile.

File: src/main/java/nwoolcan/model/brewery/warehouse/WarehouseImpl.java (hunk 2 above)
Comment (samueleceroni) [reply src/main/java/nwoolcan/model/brewery/warehouse/WarehouseImpl.java]: Please reply to the comments, don't change the comments. Thanks.

File: src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java (hunk 1 above)
Comment (samueleceroni) [reply src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java]: Are you referring to the assignment to null?

Comment (Zeegomo) [reply src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java]: yes, i'm talking about `= null`

Comment (samueleceroni) [reply src/main/java/nwoolcan/model/brewery/warehouse/QueryStockBuilder.java]: I think it is more clear and I saw it also in other codes in this repo.

**Pull request merged**